storagenode/peer.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storagenode
+
+import (
+	"context"
+	"net"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/kademlia"
+	"storj.io/storj/pkg/node"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/piecestore/psserver"
+	"storj.io/storj/pkg/piecestore/psserver/psdb"
+	"storj.io/storj/pkg/provider"
+	"storj.io/storj/pkg/server"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+// DB is the master database for Storage Node
+type DB interface {
+	// TODO: use better interfaces
+	Disk() string
+	PSDB() *psdb.DB
+	RoutingTable() (kdb, ndb storage.KeyValueStore)
+}
+
+// Config is all the configuration parameters for a Storage Node
+type Config struct {
+	Identity identity.Config
+
+	// TODO: switch to using server.Config when Identity has been removed from it
+	PublicAddress string `help:"public address to listen on" default:":7777"`
+	Kademlia      kademlia.Config
+	Piecestore    psserver.Config
+}
+
+// Verify verifies whether configuration is consistent and acceptable.
+func (config *Config) Verify() error {
+	return nil
+}
+
+// Peer is the representation of a Storage Node.
+type Peer struct {
+	// core dependencies
+	Log      *zap.Logger
+	Identity *identity.FullIdentity
+	DB       DB
+
+	// servers
+	Public struct {
+		Listener net.Listener
+		Server   *server.Server
+	}
+
+	// services and endpoints
+	RoutingTable     *kademlia.RoutingTable
+	Kademlia         *kademlia.Kademlia
+	KademliaEndpoint *node.Server
+
+	Piecestore *psserver.Server // TODO: separate into endpoint and service
+}
+
+// New creates a new Storage Node.
+func New(log *zap.Logger, full *identity.FullIdentity, db DB, config Config) (*Peer, error) {
+	peer := &Peer{
+		Log:      log,
+		Identity: full,
+		DB:       db,
+	}
+
+	var err error
+
+	{ // setup listener and server
+		peer.Public.Listener, err = net.Listen("tcp", config.PublicAddress)
+		if err != nil {
+			return nil, errs.Combine(err, peer.Close())
+		}
+
+		publicConfig := provider.ServerConfig{Address: peer.Public.Listener.Addr().String()}
```

Comment:
Why using translation than direct `server.Config`?

Reply:
Because I forgot to change it. Will fix.

Reply:
fixed